pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -1,298 +0,0 @@
-	"k8s.io/apimachinery/pkg/util/intstr"
-	"k8s.io/client-go/kubernetes"
-	corev1listers "k8s.io/client-go/listers/core/v1"
-	"k8s.io/klog/v2"
-
-	"github.com/openshift/cluster-etcd-operator/lib/resourceapply"
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/etcd_assets"
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
-)
-
-const (
-	EtcdGuardDeploymentName   = "etcd-quorum-guard"
-	infrastructureClusterName = "cluster"
-	clusterConfigName         = "cluster-config-v1"
-	clusterConfigKey          = "install-config"
-	clusterConfigNamespace    = "kube-system"
-)
-
-type replicaCountDecoder struct {
-	ControlPlane struct {
-		Replicas string `yaml:"replicas,omitempty"`
-	} `yaml:"controlPlane,omitempty"`
-}
-
-var pdb = &policyv1beta1.PodDisruptionBudget{
-	ObjectMeta: metav1.ObjectMeta{
-		Name:      EtcdGuardDeploymentName,
-		Namespace: operatorclient.TargetNamespace,
-	},
-	Spec: policyv1beta1.PodDisruptionBudgetSpec{
-		MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: int32(1)},
-		Selector:       &metav1.LabelSelector{MatchLabels: map[string]string{"k8s-app": EtcdGuardDeploymentName}},
-	},
-}
-
-// QuorumGuardController watches the etcd quorum guard deployment, create if not exists
-type QuorumGuardController struct {
-	operatorClient       v1helpers.OperatorClient
-	kubeClient           kubernetes.Interface
-	podLister            corev1listers.PodLister
-	nodeLister           corev1listers.NodeLister
-	infrastructureLister configv1listers.InfrastructureLister
-	clusterTopology      configv1.TopologyMode
-	replicaCount         int
-	etcdQuorumGuard      *appsv1.Deployment
-	cliImagePullSpec     string
-}
-
-func NewQuorumGuardController(
-	operatorClient v1helpers.OperatorClient,
-	kubeClient kubernetes.Interface,
-	kubeInformers v1helpers.KubeInformersForNamespaces,
-	eventRecorder events.Recorder,
-	infrastructureLister configv1listers.InfrastructureLister,
-	cliImagePullSpec string,
-) factory.Controller {
-	c := &QuorumGuardController{
-		operatorClient:       operatorClient,
-		kubeClient:           kubeClient,
-		podLister:            kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
-		nodeLister:           kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
-		infrastructureLister: infrastructureLister,
-		replicaCount:         0,
-		cliImagePullSpec:     cliImagePullSpec,
-	}
-	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
-		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
-		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
-		operatorClient.Informer(),
-	).WithSync(c.sync).ToController("QuorumGuardController", eventRecorder.WithComponentSuffix("quorum-guard-controller"))
-}
-
-func (c *QuorumGuardController) getTopologyMode() (configv1.TopologyMode, error) {
```

Comment:
how does the library-go controller support conditional omission? for example, we do not want QG enforcement in a single node install. I see the infra lister should this just work now?

Reply:
Correct, that has been handled [here](https://github.com/openshift/library-go/pull/1095/commits/54e116269c42eaaa86943982355ba509040794ee#diff-d0ce926df996bfa572967a0d2a1d011bbfc84cd41e065b957328285647b17cecR284)